src/utils/auto_play_harness.py

```diff
@@ -0,0 +1,177 @@
+import os
+import time
+from datetime import datetime
+from typing import List, TextIO
+
+from pydantic_yaml import parse_yaml_raw_as
+from steamship import Block, Steamship, Workspace
+from steamship.agents.schema import AgentContext
+from steamship.data import TagKind
+from steamship.data.block import StreamState
+from steamship.data.tags.tag_constants import RoleTag
+
+from api import AdventureGameService
+from generators.generator_context_utils import set_image_generator, set_music_generator
+from schema.characters import HumanCharacter
+from schema.game_state import ActiveMode
+from schema.server_settings import ServerSettings
+from utils.context_utils import (
+    get_game_state,
+    get_story_text_generator,
+    save_game_state,
+    save_server_settings,
+)
+from utils.dummy_generator import DummyGenerator
+from utils.tags import QuestArcTag, QuestTag, TagKindExtensions
+
+output_tags = [
+    (TagKindExtensions.QUEST_ARC, QuestArcTag.RESULT),
+    (TagKindExtensions.QUEST, QuestTag.QUEST_CONTENT),
+    (TagKindExtensions.QUEST, QuestTag.DICE_ROLL),
+    (TagKindExtensions.QUEST, QuestTag.ITEM_GENERATION_CONTENT),
+    (TagKindExtensions.QUEST, QuestTag.QUEST_SUMMARY),
+]
+
+
+class AutoPlayHarness:
+    server_settings: ServerSettings
+    character: HumanCharacter
+    workspace: Workspace
+    client: Steamship
+    service: AdventureGameService
+    last_seen_block = -1
+    last_content_block: Block
+    context: AgentContext
+    output_file: TextIO
+
+    def __init__(
+        self, server_settings_path: str, character_path: str, output_path: str
+    ):
+        with open(server_settings_path) as settings_file:
+            yaml_string = settings_file.read()
+            self.server_settings = parse_yaml_raw_as(ServerSettings, yaml_string)
+
+        with open(character_path) as character_file:
+            yaml_string = character_file.read()
+            self.character = parse_yaml_raw_as(HumanCharacter, yaml_string)
+
+        self.client = Steamship()
+        self.workspace = Workspace.create(self.client)
+        self.client.switch_workspace(workspace_id=self.workspace.id)
+
+        self.service = AdventureGameService(client=self.client)
+
+        self.context = self.service.build_default_context()
+        save_server_settings(self.server_settings, self.context)
+        game_state = get_game_state(self.context)
+        game_state.player = self.character
+        save_game_state(game_state, self.context)
+
+        # Override image and music generators
+        dummy_generator = DummyGenerator(self.client)
+        set_image_generator(self.context, dummy_generator)
+        set_music_generator(self.context, dummy_generator)
+
+        self.output_file = open(output_path, "w")
```

Comment:
question (non-blocking): do we need to close this at any point?  or is that not an issue?

Reply:
It is closed in the `finish` method, which I just now realized I wasn't actually calling.  Updated.